File: evidence/test116.py
"""Local-only: compile portable production helpers, inspect Objective-C wiring.
No UIKit/AVFoundation execution is claimed by this test.
"""
from pathlib import Path
import subprocess, tempfile
root=Path(__file__).resolve().parents[1]
s=(root/'src/WCCContentViewController.m').read_text()
m=(root/'src/WCCMedia.m').read_text()
h=(root/'src/WCCRuntime.h').read_text()
def method(name,next_name):
    return s.split('- (void)'+name+' {',1)[1].split('- (void)'+next_name,1)[0]
refresh=method('refreshHourlyMedia','viewDidLayoutSubviews')
assert 'WCCSafePath' not in refresh and 'stat(' not in refresh and 'WCCPrefs' not in refresh
assert 'if (item.boundIdentity && ![item.boundIdentity isEqual:identity])' in refresh
assert 'if (!item.boundIdentity)' in refresh
assert refresh.index('[item.media loadPath:nil]')<refresh.index('for (WCCHourlyItem *item in admitted)')
assert 'WCCHourlyAdmit(visible,item.cachedPath!=nil)' in refresh
assert 'WCCHourlyAnimatedLimit' not in (root/'src/WCCHourly.h').read_text()
assert '_isExpanded && self.mediaVisible && !self.mediaSuspended && self.view.window && !self.presentedViewController' in refresh
assert 'item.media.active=NO; [item.media loadPath:nil]; item.boundIdentity=nil;' in refresh
assert 'DISPATCH_QUEUE_SERIAL' in m and 'dispatch_async(WCCPreparationQueue()' in m
assert '10*NSEC_PER_SEC' in m and '[asset cancelLoading]' in m
assert m.count('generation != self->_generation')>=2
assert 'if (!current) return;' in m
for lifecycle in ['viewWillDisappear:', 'controlCenterDidDismiss', 'willResignActive']:
    body=s.split('- (void)'+lifecycle,1)[1].split('\n- (',1)[0]
    assert '[self refreshHourlyMedia]' in body
original=(root/'evidence/original104-WCCContentViewController.m').read_text()
# All original six-element expanded constraints and hourly-cell numeric frames survive.
import re
for line in s.split('- (void)layoutOriginalExpanded:',1)[1].split('- (void)refreshWeatherData',1)[0].splitlines():
    if 'constraintEqualTo' in line:
        assert line.strip().rstrip(',') in original
for frame in ['CGRectMake(12 + i * 55, 0, 55, 80)', 'CGRectMake(0, 0, 55, 18)', 'CGRectMake(12, 22, 30, 30)', 'CGRectMake(0, 56, 55, 20)']:
    assert frame in original and frame in s
assert '[self cacheHourlyMediaPaths]' in method('preferencesChanged','viewDidAppear:')
assert '[self cacheHourlyMediaPaths]' in method('updateHourlyForecast','unused')
assert 'self.greetingLabel.hidden=YES' not in s and 'self.greetingLabel.hidden=_isExpanded' not in s
expanded=s.split('- (void)layoutOriginalExpanded:',1)[1].split('- (void)refreshWeatherData',1)[0]
assert 'CGRectMake(0,0,size.width,85)' in expanded
assert 'CGRectMake(0,85,size.width,size.height-85)' in expanded
assert expanded.index('[_headerView layoutIfNeeded]')<expanded.index('occupied[n++]=WCCR')
assert 'WCCExpandedGreeting(size.width,occupied,n)' in expanded
assert 'systemFontOfSize:8' in expanded and 'NSLineBreakByTruncatingTail' in expanded
assert 'if (_active == active) return;' in m
assert 'generation != self->_generation' in m and 'WCCMediaCallbackCurrent(' in m
assert 'WCCHourlyAsset((int)code,WCCHourlyDaylight(forecast)' in s
assert 'WCCABICompatible(s.methodReturnType,"B",2,args,0,NULL)' in s
assert 'WCCABICompatible(s.methodReturnType,"c",2,args,0,NULL)' in s
assert 'NSString *key = _currentCity ? [self imageNameForConditionCode:[_currentCity conditionCode]] : nil;' in s
assert 'old.timeLabel.text=item.timeLabel.text' in s
assert 'Icon: https://i.imgs.ovh/2026/09/20/17d0953ee53152da8e61955e8bcf444a.png' in (root/'control').read_text()
with tempfile.TemporaryDirectory() as d:
    binary=Path(d)/'test116'
    subprocess.run(['cc','-std=gnu11','-Wall','-Wextra',str(root/'evidence/test116.c'),'-lm','-o',str(binary)],check=True)
    result=subprocess.run([str(binary)],capture_output=True,text=True,check=True)
    (root/'evidence/preview116-expanded.svg').write_text(result.stdout)
    # Numerically freeze all unaffected collapsed-size branches against the
    # exact current production header with only the 3x1 block reverted.
    start=h.index('        // Only 3x1:')
    end=h.index('        g.icon=WCCR(x,9*s',start)
    old=h[:start]+'        double x=p+leftWidth+6*s, tx=x+30*s+6*s;\n'+h[end:]
    old=old.replace('WCC_RUNTIME_H','OLD_RUNTIME_H').replace('WCCComputeModuleGeometry','OldModule').replace('WCCComputeGeometry','OldGeometry')
    # Separate executables avoid typedef conflicts.
    outputs=[]
    for index,header in enumerate([h,old]):
        hdr=Path(d)/f'geometry{index}.h'; hdr.write_text(header)
        program=Path(d)/'compare.c'
        fn='WCCComputeModuleGeometry' if index==0 else 'OldModule'
        program.write_text('#include "'+str(hdr)+'"\nint main(){for(int c=1;c<=4;c++)for(int r=1;r<=2;r++)for(int w=140;w<440;w+=20){if(c==3&&r==1)continue;WCCGeometry g='+fn+'(w,r*76,0,c,r); WCCRect a[]={g.icon,g.city,g.temperature,g.condition,g.highLow,g.precipitation,g.greeting};for(int i=0;i<7;i++)printf("%.6f %.6f %.6f %.6f\\n",a[i].x,a[i].y,a[i].w,a[i].h);}return 0;}')
        exe=Path(d)/f'compare{index}'
        subprocess.run(['cc',str(program),'-lm','-o',str(exe)],check=True)
        outputs.append(subprocess.check_output([str(exe)]))
    assert outputs[0]==outputs[1]
print('PASS 116: production geometry, 144 daylight mappings, 1000 admission rounds, unchanged other sizes, wiring and control Icon')

File: evidence/test116.c
#include "../src/WCCRuntime.h"
#include "../src/WCCHourly.h"
#include <assert.h>
static int overlap(WCCRect a,WCCRect b) { return a.x<b.x+b.w && a.x+a.w>b.x && a.y<b.y+b.h && a.y+a.h>b.y; }
int main(void) {
    for(int w=240;w<=420;w+=20) {
        WCCGeometry g=WCCComputeModuleGeometry(w,76,0,3,1);
        double old=12+65+6;
        assert(g.icon.x>=old && g.icon.x<=old+(w-12-old-130)*.55);
        assert(w-(g.city.x+g.city.w)>12);
        assert(g.temperature.x==12 && g.tempFont==31);
        assert(g.city.x==g.icon.x+36 && g.city.x+g.city.w<=w-12);
        assert(!overlap(g.temperature,g.icon));
        // Expanded fixtures approximate UIKit intrinsic font metrics. Production
        // uses ACTUAL resolved frames instead; fixture values are not UI proof.
        WCCRect occupied[]={WCCR(16,15,55,55),WCCR(89,12,w-170,22),WCCR(89,36,100,16),WCCR(89,54,110,15),WCCR(w-100,10,84,46),WCCR(w-110,56,94,16)};
        WCCRect r=WCCExpandedGreeting(w,occupied,6);
        assert(r.w>0 && r.h==10 && r.y+r.h<=85 && r.x>=16 && r.x+r.w<=w-16);
        for(int i=0;i<6;i++) assert(!overlap(r,occupied[i]));
        if(w==320) {
            printf("<svg xmlns='http://www.w3.org/2000/svg' width='640' height='360' viewBox='0 0 320 180'><rect width='320' height='180' fill='#203044'/><path d='M16 85H304' stroke='white'/>");
            for(int i=0;i<6;i++) printf("<rect x='%g' y='%g' width='%g' height='%g' fill='none' stroke='#88bbee'/>",occupied[i].x,occupied[i].y,occupied[i].w,occupied[i].h);
            printf("<rect x='%g' y='%g' width='%g' height='%g' fill='#408850'/><text x='%g' y='%g' fill='white' font-size='8'>Greeting (tail truncated)</text><text x='16' y='110' fill='white' font-size='10'>Hourly: unchanged y=85; header=85 total=180</text></svg>\n",r.x,r.y,r.w,r.h,r.x,r.y+8);
        }
    }
    for(int round=0;round<1000;round++) {
        int accepted=0;
        for(int i=0;i<12;i++) accepted+=WCCHourlyAdmit(1,1);
        assert(accepted==12);
        assert(WCCHourlyAdmit(1,1));
        assert(!WCCHourlyAdmit(0,1));
        assert(!WCCHourlyAdmit(1,0));
    }
    assert(!WCCHourlyIntersects(55,55,0,55));
    assert(WCCHourlyIntersects(54,55,0,55));
    char name[128],expected[128];
    for(int code=0;code<48;code++) for(int day=-1;day<=1;day++) {
        int ok=WCCHourlyAsset(code,day,name,sizeof(name));
        if(day<0 && strstr(WCCAssetTemplate(code),"%@")) assert(!ok);
        else { WCCAssetBasename(code,day==0,expected,sizeof(expected)); assert(ok && !strcmp(name,expected)); }
    }
    return 0;
}
